R/agg_helpers.R

```diff
@@ -0,0 +1,236 @@
+
+      # drop parent data already present
+      if (nrow(parent_dt) > 0) {
+        parent_dt[, drop := TRUE]
+        dt <- merge(
+          dt, parent_dt[, .SD, .SDcols = c(id_cols, "drop")],
+          all = TRUE, by = id_cols
+        )
+        dt <- dt[is.na(drop)]
+        dt[, drop := NULL]
+      }
+    }
+
+    aggregated_same_groupings_dt <- agg_subtree(
+      agg_data,
+      id_cols,
+      value_cols,
+      col_stem,
+      col_type,
+      agg_function,
+      subtree,
+      missing_dt_severity,
+      overlapping_dt_severity,
+      collapse_interval_cols
+    )
+
+    if (col_type == "interval") {
+      gen_name(aggregated_same_groupings_dt, col_stem = col_stem,
+               format = "interval")
+      setnames(aggregated_same_groupings_dt, paste0(col_stem, "_name"),
+               col_stem)
+    }
+
+    # append to final aggregated data to return
+    if (col_type == "categorical") {
+      result_dt <- rbind(result_dt, aggregated_same_groupings_dt, use.names = T)
+    } else if (col_type == "interval") {
+      result_dt[[i]] <- aggregated_same_groupings_dt
+    }
+  }
+  if (col_type == "interval") {
+    result_dt <- rbindlist(result_dt, use.names = TRUE)
+    result_dt[, c(col_stem) := NULL]
+  }
+  return(result_dt)
+}
+
+#' @title Helper function to aggregate interval variables
+#'
+#' @description Helper function to aggregate interval variables by groups of data
+#'   that have the same intervals. Calls `agg_categorical()` for each group.
+#'
+#' @inheritParams agg
+#'
+#' @return Invisibly returns `dt` but throws error if inputs are not formatted
+#'   correctly.
+agg_interval <- function(dt,
+                         id_cols,
+                         value_cols,
+                         col_stem,
+                         col_type,
+                         mapping,
+                         agg_function = sum,
+                         missing_dt_severity = "stop",
+                         present_agg_severity = "stop",
+                         overlapping_dt_severity = "stop",
+                         na_value_severity = "stop",
+                         collapse_interval_cols = FALSE,
+                         quiet = FALSE) {
+
+  cols <- paste0(col_stem, "_", c("start", "end"))
+
+  # create one column to describe each interval
+  gen_name(mapping, col_stem = col_stem, format = "interval")
+  data.table::setnames(mapping, paste0(col_stem, "_name"), col_stem)
+
+  # create one column to describe each interval
+  dt_intervals <- unique(dt[, .SD, .SDcols = cols])
+  gen_name(dt_intervals, col_stem = col_stem, format = "interval")
+  data.table::setnames(dt_intervals, paste0(col_stem, "_name"), col_stem)
+  data.table::setkeyv(dt_intervals, cols)
+  dt <- dt[dt_intervals, on = cols, nomatch = 0]
+
+  groups <- dt[, list(col = paste(get(col_stem), collapse = ",")), by = setdiff(names(dt), c(col_stem, cols, value_cols))]
+  unique_groups <- unique(groups$col)
```

Comment:
identifies unique groupings of interval variables to aggregate all at once.

For example aggregates all rows with only [0, Inf] interval, then aggregates all rows with 5 year age groups up to 95+, etc.